### PR TITLE
Allow After/BeforeTestWithContext usage in superclass

### PR DESCRIPTION
**Pull request by aczajkowski**
Allowing usage of
- @BeforeTestWithContext 
- @AfterTestWithContext 

in superclasses. It's often used to build a common set of tests and their initialisation for multiple contexts.

### CODE REVIEW

**Comment by aczajkowski** [comment]
CC: @findepi

**Comment by findepi** [comment]
Would it be possible to have a test for that?


Also, let's make sure (i.e. have a test coverage) than an overridden setup method gets invoked once.

- when super is annotated
- when super and overriding are annotated

**Comment by aczajkowski** [comment]
> Would it be possible to have a test for that?
> 
> Also, let's make sure (i.e. have a test coverage) than an overridden setup method gets invoked once.
> 
> * when super is annotated
> * when super and overriding are annotated

@findepi  Yes added.

**Review by findepi — APPROVED** [review]

File: tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy
Comment (kokosing) [inline tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy:227]: Do we need to annotate it once again? Why not to take annotation from `TestClass`? Does it change the order in which methods are invoked? Consider a case:

```
class B { @BeforeTestWithContext void b() {} }
class A extends B { @BeforeTestWithContext void a() {} }
```

vs 

```
class B { @BeforeTestWithContext void b() {} }
class A extends B { @BeforeTestWithContext void a() {} void b() {} }
```

Will `b()` be called always as first?

Comment (aczajkowski) [reply tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy:227]: Yes b() will be always called first, as it will be always returned first from getMethods() call.
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getMethods()
I will add a test to call annotated methods on multiple levels.

Comment (kokosing) [reply tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy:227]: Should we have test for that?

Comment (aczajkowski) [reply tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy:227]: Just added: TestClassOverrideAdditionalAnnotatedMethod 😄  https://github.com/aczajkowski/tempto/blob/cbc2021ea3353cf86e6b60573d864f7b4a2b067d/tempto-core/src/test/groovy/io/trino/tempto/internal/initialization/TestInitializationListenerTest.groovy#L115

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
I don't see CI results.

**Comment by findepi** [comment]
@aczajkowski can you please push an `empty` commit to force CI run?

**Comment by aczajkowski** [comment]
> @aczajkowski can you please push an `empty` commit to force CI run?

Seems CI is not starting. No idea why 🤷‍♂️

**Comment by hashhar** [comment]
Seems something wrong with Travis CI - https://travis-ci.com/github/trinodb/tempto/requests.

**Comment by findepi** [comment]
i wasn't aware we're still on travis with this one. Maybe that's why. Travis UI tells me that

> Builds have been temporarily disabled for private and public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance.

**Comment by hashhar** [comment]
We might need to pick up https://github.com/trinodb/tempto/pull/62 and drive it to completion first.
I can take a look after-hours today if this is not very urgent.

**Comment by aczajkowski** [comment]
@findepi @hashhar Hi I've got back after holidays, how can we follow up on this ?

**Comment by hashhar** [comment]
I've created https://github.com/trinodb/tempto/pull/83 as first part of moving to GHA. I have another branch with actual changes which I will open a PR from once #83 gets merged (since workflows don't run unless they are present in master branch).

Once that is done this branch can be tested and we can proceed.

**Comment by hashhar** [comment]
@aczajkowski CI is now working again (some tests are already failing on master). I think you should be fine to proceed with this now.

**Comment by aczajkowski** [comment]
@hashhar Tests passed.
Only exclusion test on Java 11 and 15 failed. I've pushed a commit to fix that. It was simply wrong package name in GHA workflow task.
```
-e io.prestosql.tempto.examples.ExclusionTest.failingTest,io.prestosql.tempto.examples.AnotherExclusionTest
```
And now our package is:
```io.trino....```

**Comment by aczajkowski** [comment]
@hashhar @findepi As I'm fresh here, How we get merge. I have no privileges, green build but 2 approvals :)

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**
